types/ckeditor__ckeditor5-core/src/plugincollection.d.ts

```diff
@@ -15,7 +15,7 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
 
     [Symbol.iterator](): Iterator<[typeof Plugin, Plugin]>;
     destroy(): Promise<void>;
-    get(key: (() => Plugin) | string): Plugin;
+    get(key: (() => Plugin) | string): new () => Plugin;
```

Comment:
dbl checkin, you wanted `constructable` here, hence `new`?

Reply:
Since `Plugin` is an abstract class, it seems that I have to write it this way to say that it will return an instance of a class that extends said `Plugin` class.

Reply:
How are you using this? This has broken my usage -- I was previously doing

```ts
let somePlugin = editor.plugins.get(SomePlugin.pluginName) as SomePlugin
```

which was working. With this change, ts complains that the types don't overlap so I have to first cast to `unknown`. This change looks to me like it's switching the return type from a `Plugin` instance to a `Plugin` class/constructor, which isn't correct.

Am I misunderstanding? How are you invoking `editor.plugins.get()`?

Reply:
From my experimentation, it looks like we want something like this:

```ts
type PluginConstructor<T = Plugin> = { new (...args: unknown[]): T }

export default class PluginCollection implements Emitter, Iterable<[typeof Plugin, Plugin]> {
  // ...
  get<T extends Plugin>(key: PluginConstructor<T>): T;
  get(key: string): Plugin;
  has<T extends Plugin>(key: PluginConstructor<T>): boolean;
  has(key: string): boolean;
  // ...
}
```

This allows both

```ts
let somePlugin = editor.plugins.get(SomePlugin);
```

and

```ts
let somePlugin = editor.plugins.get(SomePlugin.pluginName) as SomePlugin
```

and the equivalents for `.has()`.

Thoughts?

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54924